### ISSUE TITLE
AQL: cannot instantiate CollectBlock with undetermined aggregation method

### ISSUE THREAD

**Issue by ThomasWeiss** [arangodb/arangodb#3395]
Running 3.2.4 with RocksDB on Ubuntu 16.04 single-server.

I got a weird error when executing the following query (shortened to highlight the problem):
```
WITH keaks, battles, beats, users
LET likedKeaks = (
    FOR e IN events_ 
    FILTER e.type == 'keakLikeCreated' && e.timestamp > '2017-10-09' 
    COLLECT keakId = CONCAT('keaks/', e.data.keakId) WITH COUNT INTO length
    RETURN { id: keakId, score: length * 3 })
LET likedBattles = (
    FOR e IN events_ 
    FILTER e.type == 'battleLikeCreated' && e.timestamp > '2017-10-09' 
    COLLECT battleId = CONCAT('battles/', e.data.battleId) WITH COUNT INTO length
    RETURN { id: battleId, score: length * 3 })
LET commentedKeaks = (
    FOR e IN events_ 
    FILTER e.type == 'keakCommentCreated' && e.timestamp > '2017-10-09' 
    COLLECT keakId = CONCAT('keaks/', e.data.keakId) WITH COUNT INTO length
    RETURN { id: keakId, score: length * 2 })
LET commentedBattles = (
    FOR e IN events_ 
    FILTER e.type == 'battleCommentCreated' && e.timestamp > '2017-10-09' 
    COLLECT battleId = CONCAT('battles/', e.data.battleId) WITH COUNT INTO length
    RETURN { id: battleId, score: length * 2 })
LET sharedKeaks = (
    FOR e IN events_ 
    FILTER e.type == 'keakShared' && e.timestamp > '2017-10-09' 
    COLLECT keakId = CONCAT('keaks/', e.entityId) WITH COUNT INTO length
    RETURN { id: keakId, score: length })
LET sharedBattles = (
    FOR e IN events_ 
    FILTER e.type == 'battleShared' && e.timestamp > '2017-10-09' 
    COLLECT battleId = CONCAT('battles/', e.entityId) WITH COUNT INTO length
    RETURN { id: battleId, score: length })
LET viewedKeaks = (
    FOR e IN events_ 
    FILTER e.type == 'keakViewsRecorded' && e.timestamp > '2017-10-09' 
    COLLECT keakId = CONCAT('keaks/', e.entityId) AGGREGATE length = SUM(e.data.viewCount)
    RETURN { id: keakId, score: length })
LET popularContent = UNION(likedKeaks, likedBattles, commentedKeaks, commentedBattles, sharedKeaks, sharedBattles, viewedKeaks)
FOR content IN popularContent
COLLECT contentId = content.id
AGGREGATE score = SUM(content.score)
RETURN { id: contentId, score: score }
```

This fails with: `Query: AQL: cannot instantiate CollectBlock with undetermined aggregation method (exception location: /var/lib/otherjenkins/workspace/RELEASE__BuildPackages/arangod/Aql/ExecutionEngine.cpp:143). Please report this error to arangodb.com (while executing) (exception location: /var/lib/otherjenkins/workspace/RELEASE__BuildPackages/arangod/RestHandler/RestCursorHandler.cpp:130). Please report this error to arangodb.com`

Now if I remove **any** one of the seven sub-queries, it works. I only got the issue when all seven sub-queries are declared and included in the `UNION`. Let me know if you need further information.

**Comment by jsteemann** [comment]
It looks like the optimizer generates a lot of possible execution plans until it reaches the default built-in cap.
After that, a required optimizer rule does not seem to be applied anymore, which then causes the error.
Potential fix:
```diff
diff --git a/arangod/Aql/Optimizer.cpp b/arangod/Aql/Optimizer.cpp
index dd3a2f6842..942c2db6bb 100644
--- a/arangod/Aql/Optimizer.cpp
+++ b/arangod/Aql/Optimizer.cpp
@@ -132,7 +132,7 @@ int Optimizer::createPlans(ExecutionPlan* plan,
         // skip over rules if we should
         // however, we don't want to skip those rules that will not create
         // additional plans
-        if ((runOnlyRequiredRules && rule.canCreateAdditionalPlans) ||
+        if ((runOnlyRequiredRules && rule.canCreateAdditionalPlans && rule.canBeDisabled) ||
             _disabledIds.find(level) != _disabledIds.end()) {
           // we picked a disabled rule or we have reached the max number of
           // plans and just skip this rule
```

**Comment by jsteemann** [comment]
This has been fixed in 3.2.5, which has been released meanwhile.
